File: 프로그래머스/2/77485. 행렬 테두리 회전하기/행렬 테두리 회전하기.py
def rotate(pos, maps):
    dx, dy = [0, 1, 0, -1], [1, 0, -1, 0]
    sx, sy, ex, ey = pos
    x, y, d = sx, sy, 0
    next = maps[sx][sy]
    ans = maps[sx][sy]
    
    while True:
        nx, ny = x + dx[d], y + dy[d]
        if sx <= nx <= ex and sy <= ny <= ey:
            ans = min(ans ,maps[nx][ny])
            maps[nx][ny], next = next, maps[nx][ny]
        else:
            d += 1
            nx, ny = x + dx[d], y + dy[d]
            ans = min(ans ,maps[nx][ny])
            maps[nx][ny], next = next, maps[nx][ny]
        if nx == sx and ny == sy: break
        x, y = nx, ny   

    return ans
def solution(rows, columns, queries):
    maps = [[(i*columns + (j+1)) for j in range(columns)] for i in range(rows)]
    return [rotate([x-1 for x in arr], maps) for arr in queries]
"""
테두리 부분 시계방향으로 한칸씩 회전
100 * 100 크기 맵
회전 10000개
x행 y열
1. 모든 회전에 대해 다음을 반복
    1-1. 경계에 도착하면 방향전환
    1-2. 최소숫자찾기
    1-3. 회전 완료되면 숫자 저장
2. 저장된 숫자 반환
"""

File: 프로그래머스/2/77485. 행렬 테두리 회전하기/행렬 테두리 회전하기.java
import java.util.*;
class Solution {
    static int[][] board;
    static int[] dx = {0, 1, 0, -1}, dy = {1, 0, -1, 0};
    private int solve(int[] condi){
        for(int i = 0; i < condi.length;i++){
            condi[i]--;
        }
        int x1= condi[0], y1 = condi[1], x2= condi[2], y2 = condi[3];
        int[][] temp = new int[x2-x1+1][y2-y1+1];
        int x = x1, y = y1, d = 0;
        int nx = x + dx[d], ny = y + dy[d];
        temp[nx-x1][ny-y1] = board[x][y];
        x = nx; y = ny;
        while(x != x1 || y != y1){
            nx = x + dx[d]; ny = y + dy[d];
            if(nx < x1 || ny < y1 || nx > x2 || ny > y2){
                d = (d + 1) % 4;
                nx = x + dx[d]; ny = y + dy[d];
            };
            temp[nx - x1][ny - y1] = board[x][y];
            x = nx; y = ny;
        }
        int ret = 1000 * 1000;
        for(int i = x1; i <= x2; i++){
            for(int j = y1; j <= y2; j++){
                int now = temp[i-x1][j-y1];
                if(now == 0) continue;
                ret = Math.min(ret, now);
                board[i][j] = now;
            }
        }
        return ret;
    }
    public List<Integer> solution(int r, int c, int[][] queries) {
        board = new int[r][c];
        int num = 1;
        for(int i = 0; i < r; i++){
            for(int j = 0; j < c; j++){
                board[i][j] = num++;
            }
        }
        var ret = new ArrayList<Integer>();
        for(int[] condi : queries){
            ret.add(solve(condi));
        }
        return ret;
    }
}
/*
System.out.println(Arrays.toString(arr));
1~ r*c 숫자 적힘
직사각형 테두리에 있는 숫자를 시계방향으로 한칸씩밀기
회전에 의해 바뀐 숫자들 중 가장 작은 숫자들을 배열에 담아 반환한다

1만번의 회전?? 최대 400만 가능 
회전을 어떻게 표현할까??
1. 크기만큼의 배열을 만든다 
2. 원본배열에서 회전을 하면서 새로운 배열의 직사각형을 채운다
3. 새로운 배열의 직사각형을 돌면서 원본배열의 크기를 채운다 
4. 회전이 끝난후 가장작은 숫자를 반환 한다


*/